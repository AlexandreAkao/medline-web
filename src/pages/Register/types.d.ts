type UserRegisterTypes =
  | 'name'
  | 'password'
  | 'email'
  | 'birthdate'
  | 'cpf'
  | 'rg'
  | 'cep'
  | 'street'
  | 'number'
  | 'city'
  | 'state'
  | 'district';

interface IUserRegisterAction {
  type: UserRegisterTypes;
  payload: string;
}
