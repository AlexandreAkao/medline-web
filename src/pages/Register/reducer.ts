export const userFormReducer = (state: IUserRegister, action: IUserRegisterAction) => {
  switch (action.type) {
    case 'name':
    case 'password':
    case 'email':
    case 'cpf':
    case 'rg':
      return { ...state, [action.type]: action.payload };
    case 'birthdate':
      return { ...state, [action.type]: action.payload.split('/').reverse().join('-') };
    case 'cep':
    case 'street':
    case 'number':
    case 'city':
    case 'state':
    case 'district':
      return { ...state, address: { ...state.address, [action.type]: action.payload } };
    default:
      return state;
  }
};

export const userFormInitialState = {
  name: '',
  password: '',
  email: '',
  birthdate: '',
  cpf: '',
  rg: '',
  address: {
    cep: '',
    city: '',
    district: '',
    number: '',
    state: '',
    street: '',
  },
};
