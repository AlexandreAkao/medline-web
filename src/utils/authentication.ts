import { isEmpty } from 'ramda';

export const getUserInfo = () => {
  const user = JSON.parse(localStorage.getItem('user') ?? '{}');

  return isEmpty(user) ? null : (user as IUser);
};

export const geToken = () => localStorage.getItem('token');
