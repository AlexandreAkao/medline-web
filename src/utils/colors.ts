export const getColorAlpha = (color: string, alpha = 1) => {
  const alphaLevel = Math.round(255 * alpha).toString(16);
  return color + alphaLevel;
};
