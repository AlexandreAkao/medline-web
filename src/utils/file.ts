export const downloadFile = (blob: Blob, filename: string) => {
  const link = document.createElement('a');
  link.href = URL.createObjectURL(blob);
  link.download = filename;
  link.click();
};
