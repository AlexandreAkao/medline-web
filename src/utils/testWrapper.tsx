import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithRouter = (component: React.ReactNode) =>
  render(<MemoryRouter>{component}</MemoryRouter>);
