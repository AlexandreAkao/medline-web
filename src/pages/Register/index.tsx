import { ChangeEvent, useReducer, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { toast } from 'react-toastify';

import {
  RegisterContainer,
  RegisterContent,
  RegisterBackground,
  RegisterFormContainer,
  RegisterFormRow,
  AddressFormSection,
  AddressFormTitle,
  RegisterFormTittle,
  AddressFormButtonsContainer,
} from 'pages/Register/styles';
import TextInput from 'components/TextInput';
import { Mask } from 'utils/mask';
import Button from 'components/Button';
import api from 'service/api';
import MedlineHeader from 'components/Header/MedlineHeader';
import { userFormInitialState, userFormReducer } from 'pages/Register/reducer';
import { useAuth } from 'hooks/useAuth';

function Register() {
  const [userForm, dispatchUserForm] = useReducer(userFormReducer, userFormInitialState);
  const [confirmPassword, setConfirmPassword] = useState('');
  const navigate = useNavigate();
  const { handleLogin } = useAuth();

  const handleGoBack = () => {
    navigate(-1);
  };

  const handleChangeInput = (type: UserRegisterTypes) => (event: ChangeEvent<HTMLInputElement>) =>
    dispatchUserForm({ payload: event.target.value, type });

  const handleRegisterUserButton = async () => {
    if (
      userForm.password !== confirmPassword ||
      userForm.password === '' ||
      confirmPassword === ''
    ) {
      toast.error('As senhas precisam ser iguais');
      return;
    }
    const data = await api.post('user', userForm);

    if (data.status === 200) {
      toast.success('Usuário criado com sucesso!');
      handleLogin({
        email: userForm.email,
        password: userForm.password,
      });
      navigate('/');
    }
  };

  return (
    <RegisterContainer>
      <MedlineHeader />

      <RegisterContent>
        <RegisterBackground />

        <RegisterFormContainer>
          <RegisterFormTittle>Informe seus dados</RegisterFormTittle>
          <RegisterFormRow>
            <TextInput
              label="Nome"
              placeholder="Nome"
              name="name"
              onChange={handleChangeInput('name')}
              value={userForm.name}
            />
            <TextInput
              label="Data de Nascimento"
              placeholder="Data de Nascimento"
              name="birth"
              mask={Mask.Date}
              onChange={handleChangeInput('birthdate')}
              value={userForm.birthdate.split('-').reverse().join('/')}
            />
          </RegisterFormRow>

          <RegisterFormRow>
            <TextInput
              label="CPF"
              placeholder="CPF"
              mask={Mask.CPF}
              onChange={handleChangeInput('cpf')}
              value={userForm.cpf}
            />
            <TextInput
              label="RG"
              placeholder="RG"
              mask={Mask.RG}
              onChange={handleChangeInput('rg')}
              value={userForm.rg}
            />
          </RegisterFormRow>
          <TextInput
            label="Email"
            placeholder="Email"
            onChange={handleChangeInput('email')}
            value={userForm.email}
          />
          <TextInput
            label="Senha"
            placeholder="Senha"
            onChange={handleChangeInput('password')}
            value={userForm.password}
            isPassword
          />
          <TextInput
            label="Confirmar senha"
            placeholder="Confirmar senha"
            onChange={e => setConfirmPassword(e.target.value)}
            value={confirmPassword}
            isPassword
          />

          <AddressFormSection>
            <AddressFormTitle>Endereço</AddressFormTitle>
            <RegisterFormRow>
              <TextInput
                label="CEP"
                placeholder="CEP"
                mask={Mask.CEP}
                onChange={handleChangeInput('cep')}
                value={userForm.address.cep}
              />
              <TextInput
                label="Estado"
                placeholder="Estado"
                onChange={handleChangeInput('state')}
                value={userForm.address.state}
              />
            </RegisterFormRow>
            <RegisterFormRow>
              <TextInput
                label="Rua"
                placeholder="Cidade"
                onChange={handleChangeInput('street')}
                value={userForm.address.street}
                name="street-address"
              />
              <TextInput
                label="Cidade"
                placeholder="Cidade"
                onChange={handleChangeInput('city')}
                value={userForm.address.city}
              />
            </RegisterFormRow>
            <RegisterFormRow>
              <TextInput
                label="Nº Residência"
                placeholder="Nº Residência"
                onChange={handleChangeInput('number')}
                value={userForm.address.number}
              />
              <TextInput
                label="Bairro"
                placeholder="Bairro"
                onChange={handleChangeInput('district')}
                value={userForm.address.district}
              />
            </RegisterFormRow>
          </AddressFormSection>
          <AddressFormButtonsContainer>
            <Button primary={false} size="large" onClick={handleGoBack}>
              Voltar
            </Button>
            <Button primary={false} size="large" onClick={handleRegisterUserButton}>
              Cadastrar
            </Button>
          </AddressFormButtonsContainer>
        </RegisterFormContainer>
      </RegisterContent>
    </RegisterContainer>
  );
}

export default Register;
