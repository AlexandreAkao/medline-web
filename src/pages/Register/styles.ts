import styled from 'styled-components';

import colors from 'styles/colors';
import doctors from 'assets/images/doctors.svg';
import ScreenSizes from 'styles/screenSizes';

export const RegisterContainer = styled.div`
  background-color: ${colors.primary.normal};
  height: 100%;
  overflow: auto;
`;

export const RegisterContent = styled.div`
  display: flex;

  > * {
    flex: 1;
    justify-content: center;
  }

  @media (max-width: ${ScreenSizes.xLarge}) {
    flex-direction: column;

    > * {
      flex: auto;
      justify-content: center;
    }
  }
`;

export const RegisterBackground = styled.div`
  background-image: url(${doctors});
  background-repeat: no-repeat;
  background-position: center;
  height: 600px;

  @media (max-width: ${ScreenSizes.xLarge}) {
    height: 300px;
    background-size: contain;
    display: none;
  }
`;

export const RegisterFormContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 0 32px 32px 32px;
  gap: 10px;
`;

export const RegisterFormTittle = styled.h1`
  color: ${colors.white.normal};
  font-weight: 400;
  font-size: 3rem;
  margin-bottom: 10px;
`;

export const RegisterFormRow = styled.div`
  display: flex;
  width: 100%;
  gap: 40px;
`;

export const AddressFormTitle = styled.h1`
  color: ${colors.white.normal};
  font-size: 1.5rem;
  position: relative;
  margin-bottom: 20px;

  ::after {
    margin-left: 8px;
    top: 15px;
    content: '';
    position: absolute;
    width: 78%;
    height: 1px;
    background-color: ${colors.white.normal};
  }
`;

export const AddressFormSection = styled.div`
  display: flex;
  flex-direction: column;
  margin: 20px 0;
  width: 100%;
  gap: 10px;
`;

export const AddressFormButtonsContainer = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-around;
`;
