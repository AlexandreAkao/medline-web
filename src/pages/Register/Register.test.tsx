import { screen } from '@testing-library/react';

import { renderWithProvider } from 'utils/testWrapper';
import Register from 'pages/Register';

describe('Register', () => {
  const renderComponent = () => {
    return renderWithProvider(<Register />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render correctly', () => {
    renderComponent();

    const titleElement = screen.getByText('Informe seus dados');
    const buttonElements = screen.getAllByRole('button');
    const inputElements = screen.getAllByRole('textbox');

    expect(titleElement).toBeInTheDocument();
    expect(inputElements).toHaveLength(13);
    expect(buttonElements).toHaveLength(3);
  });
});
