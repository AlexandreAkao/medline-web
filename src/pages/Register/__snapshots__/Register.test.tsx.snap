// Vitest Snapshot v1

exports[`Register > should match with snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ fLfyXd"
  >
    <header
      class="sc-breuTD bvHIRC"
      data-testid="test-navigation"
    >
      <img
        alt="Medline logo"
        class="sc-ksZaOG hdhcOY"
        src="/src/assets/images/medline-logo.svg"
      />
      <nav
        class="sc-hAZoDl jYIcRX"
        data-testid="test-navigation-items"
      >
        <div
          class="sc-evZas gLFSKm"
        >
          <button
            class="sc-crXcEl CHpSV"
          >
            Cadastro
          </button>
        </div>
        <div
          class="sc-evZas gLFSKm"
        >
          <a
            href="/login"
          >
            Entrar
          </a>
        </div>
      </nav>
    </header>
    <div
      class="sc-gsnTZi dXGIgA"
    >
      <div
        class="sc-dkzDqf iIpbcV"
      />
      <div
        class="sc-hKMtZM gDVjvv"
      >
        <h1
          class="sc-eCYdqJ eVEZxl"
        >
          Informe seus dados
        </h1>
        <div
          class="sc-jSMfEi dnSuBL"
        >
          <div
            class="sc-papXJ dIxUuV"
          >
            <span
              class="sc-jqUVSM dcSCwQ"
              data-testid="test-input-label"
            >
              Nome
            </span>
            <div
              class="sc-kDDrLX hotFGy"
            >
              <input
                class="sc-iqcoie jqtTIz"
                data-testid="test-input"
                name="name"
                placeholder="Nome"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="sc-papXJ dIxUuV"
          >
            <span
              class="sc-jqUVSM dcSCwQ"
              data-testid="test-input-label"
            >
              Data de Nascimento
            </span>
            <div
              class="sc-kDDrLX hotFGy"
            >
              <input
                class="sc-iqcoie jqtTIz"
                data-testid="test-input"
                name="birth"
                placeholder="Data de Nascimento"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="sc-jSMfEi dnSuBL"
        >
          <div
            class="sc-papXJ dIxUuV"
          >
            <span
              class="sc-jqUVSM dcSCwQ"
              data-testid="test-input-label"
            >
              CPF
            </span>
            <div
              class="sc-kDDrLX hotFGy"
            >
              <input
                class="sc-iqcoie jqtTIz"
                data-testid="test-input"
                placeholder="CPF"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="sc-papXJ dIxUuV"
          >
            <span
              class="sc-jqUVSM dcSCwQ"
              data-testid="test-input-label"
            >
              RG
            </span>
            <div
              class="sc-kDDrLX hotFGy"
            >
              <input
                class="sc-iqcoie jqtTIz"
                data-testid="test-input"
                placeholder="RG"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="sc-papXJ dIxUuV"
        >
          <span
            class="sc-jqUVSM dcSCwQ"
            data-testid="test-input-label"
          >
            Email
          </span>
          <div
            class="sc-kDDrLX hotFGy"
          >
            <input
              class="sc-iqcoie jqtTIz"
              data-testid="test-input"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="sc-papXJ dIxUuV"
        >
          <span
            class="sc-jqUVSM dcSCwQ"
            data-testid="test-input-label"
          >
            Senha
          </span>
          <div
            class="sc-kDDrLX hotFGy"
          >
            <input
              class="sc-iqcoie jqtTIz"
              data-testid="test-input"
              placeholder="Senha"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="sc-papXJ dIxUuV"
        >
          <span
            class="sc-jqUVSM dcSCwQ"
            data-testid="test-input-label"
          >
            Confirmar senha
          </span>
          <div
            class="sc-kDDrLX hotFGy"
          >
            <input
              class="sc-iqcoie jqtTIz"
              data-testid="test-input"
              placeholder="Confirmar senha"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="sc-iBkjds cqMmdU"
        >
          <h1
            class="sc-gKXOVf kEyzby"
          >
            Endereço
          </h1>
          <div
            class="sc-jSMfEi dnSuBL"
          >
            <div
              class="sc-papXJ dIxUuV"
            >
              <span
                class="sc-jqUVSM dcSCwQ"
                data-testid="test-input-label"
              >
                CEP
              </span>
              <div
                class="sc-kDDrLX hotFGy"
              >
                <input
                  class="sc-iqcoie jqtTIz"
                  data-testid="test-input"
                  placeholder="CEP"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="sc-papXJ dIxUuV"
            >
              <span
                class="sc-jqUVSM dcSCwQ"
                data-testid="test-input-label"
              >
                Estado
              </span>
              <div
                class="sc-kDDrLX hotFGy"
              >
                <input
                  class="sc-iqcoie jqtTIz"
                  data-testid="test-input"
                  placeholder="Estado"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="sc-jSMfEi dnSuBL"
          >
            <div
              class="sc-papXJ dIxUuV"
            >
              <span
                class="sc-jqUVSM dcSCwQ"
                data-testid="test-input-label"
              >
                Rua
              </span>
              <div
                class="sc-kDDrLX hotFGy"
              >
                <input
                  class="sc-iqcoie jqtTIz"
                  data-testid="test-input"
                  name="street-address"
                  placeholder="Cidade"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="sc-papXJ dIxUuV"
            >
              <span
                class="sc-jqUVSM dcSCwQ"
                data-testid="test-input-label"
              >
                Cidade
              </span>
              <div
                class="sc-kDDrLX hotFGy"
              >
                <input
                  class="sc-iqcoie jqtTIz"
                  data-testid="test-input"
                  placeholder="Cidade"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="sc-jSMfEi dnSuBL"
          >
            <div
              class="sc-papXJ dIxUuV"
            >
              <span
                class="sc-jqUVSM dcSCwQ"
                data-testid="test-input-label"
              >
                Nº Residência
              </span>
              <div
                class="sc-kDDrLX hotFGy"
              >
                <input
                  class="sc-iqcoie jqtTIz"
                  data-testid="test-input"
                  placeholder="Nº Residência"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="sc-papXJ dIxUuV"
            >
              <span
                class="sc-jqUVSM dcSCwQ"
                data-testid="test-input-label"
              >
                Bairro
              </span>
              <div
                class="sc-kDDrLX hotFGy"
              >
                <input
                  class="sc-iqcoie jqtTIz"
                  data-testid="test-input"
                  placeholder="Bairro"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-ftvSup ioFdsF"
        >
          <button
            class="sc-crXcEl kULOXI"
          >
            Voltar
          </button>
          <button
            class="sc-crXcEl kULOXI"
          >
            Cadastrar
          </button>
        </div>
      </div>
    </div>
  </div>
  ,
</div>
`;
